eppo_client/eval.py

```diff
@@ -0,0 +1,93 @@
+from typing import Dict, Optional, Union
+from eppo_client.sharding import Sharder
+from eppo_client.models import Flag, Range, Shard, Variation
+from eppo_client.rules import matches_rule
+from dataclasses import dataclass
+import datetime
+
+
+@dataclass
+class FlagEvaluation:
+    flag_key: str
+    subject_key: str
+    subject_attributes: Dict[str, Union[str, float, int, bool]]
+    allocation_key: Optional[str]
+    variation: Optional[Variation]
+    extra_logging: Dict[str, str]
+    do_log: bool
+
+
+@dataclass
+class Evaluator:
+    sharder: Sharder
+
+    def evaluate_flag(
+        self,
+        flag: Flag,
+        subject_key: str,
+        subject_attributes: Dict[str, Union[str, float, int, bool]],
+    ) -> Optional[FlagEvaluation]:
+        if not flag.enabled:
+            return none_result(flag.key, subject_key, subject_attributes)
+
+        now = utcnow()
+        for allocation in flag.allocations:
+            # Skip allocations that are not active
+            if allocation.start_at and now < allocation.start_at:
+                continue
+            if allocation.end_at and now > allocation.end_at:
+                continue
+
+            # Skip allocations when none of the rules match
+            # So we look for (rule 1) OR (rule 2) OR (rule 3) etc.
+            # If there are no rules, then we always match
+            if not allocation.rules or any(
+                matches_rule(rule, {"id": subject_key, **subject_attributes})
```

Comment:
This may be controversial: injecting the subject_key as "id" in the subject attributes dictionary so we can match on ID without having to explicitly put the ID in subject attributes. Note that a custom ID in subject_attributes will overwrite this ID

Reply:
Not super controversial because I'm recalling that customers have asked to be able to match on the subject key in the rules. I think it's fine as long as we mention this behavior in the docs.